File: FrontOne.Shared/Security/CryptoService.cs
using System.Security.Cryptography;
using System.Text;

namespace FrontOne.Shared.Security;

public class CryptoService : ICryptoService
{
    private const string Passphrase = "FrontOne-ERP-Empacadora-Aguacate-2026";
    private const string Salt = "FrontOne-Salt-Fijo-v1";
    private const int KeySizeBits = 256;
    private const int Iterations = 100_000;

    public string Encrypt(string plainText)
    {
        using var aes = CreateAes();
        using var encryptor = aes.CreateEncryptor();
        var plainBytes = Encoding.UTF8.GetBytes(plainText);
        var cipherBytes = encryptor.TransformFinalBlock(plainBytes, 0, plainBytes.Length);
        return Convert.ToBase64String(cipherBytes);
    }

    public string Decrypt(string cipherText)
    {
        using var aes = CreateAes();
        using var decryptor = aes.CreateDecryptor();
        var cipherBytes = Convert.FromBase64String(cipherText);
        var plainBytes = decryptor.TransformFinalBlock(cipherBytes, 0, cipherBytes.Length);
        return Encoding.UTF8.GetString(plainBytes);
    }

    private static Aes CreateAes()
    {
        var passphraseBytes = Encoding.UTF8.GetBytes(Passphrase);
        var saltBytes = Encoding.UTF8.GetBytes(Salt);
        var derivedBytes = Rfc2898DeriveBytes.Pbkdf2(passphraseBytes, saltBytes, Iterations, HashAlgorithmName.SHA256, (KeySizeBits / 8) + 16);

        var aes = Aes.Create();
        aes.KeySize = KeySizeBits;
        aes.Key = derivedBytes[..(KeySizeBits / 8)];
        aes.IV = derivedBytes[(KeySizeBits / 8)..];
        aes.Mode = CipherMode.CBC;
        aes.Padding = PaddingMode.PKCS7;
        return aes;
    }
}

File: FrontOne.Android/domain/src/main/kotlin/com/frontone/android/domain/crypto/CryptoService.kt
package com.frontone.android.domain.crypto

import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Réplica EXACTA de FrontOne.Shared/Security/CryptoService.cs (FrontOne WinForms) —
 * misma passphrase/salt/iteraciones/derivación de clave e IV. Necesario para poder
 * descifrar Seguridad.Usuario.PasswordEncriptado tal como lo guardó el escritorio
 * (es cifrado AES simétrico reversible, no un hash de una vía).
 *
 * DECISIÓN DE SEGURIDAD DOCUMENTADA (confirmada con el usuario, 2026-08-07): esto
 * empotra la misma clave maestra en el APK que ya vive hardcodeada en el ensamblado
 * de FrontOne.WinForms.exe — no es un riesgo nuevo que introduce Android, es el mismo
 * riesgo que ya existe hoy en el ejecutable de escritorio (ambos son igual de fáciles
 * de decompilar). Se decidió portar igual que escritorio por consistencia, en vez de
 * mover la validación a un SP del lado servidor (que hoy tampoco existe en WinForms).
 *
 * Únicamente `decrypt` está implementado — Android nunca necesita cifrar contraseñas
 * (eso solo pasa al dar de alta/editar un Usuario, función que sigue siendo exclusiva
 * de escritorio).
 *
 * Nunca cambiar estas constantes sin cambiar también las de CryptoService.cs — deben
 * quedar bit a bit idénticas o ningún password existente se podrá descifrar.
 */
object CryptoService {
    private const val PASSPHRASE = "FrontOne-ERP-Empacadora-Aguacate-2026"
    private const val SALT = "FrontOne-Salt-Fijo-v1"
    private const val KEY_SIZE_BITS = 256
    private const val ITERATIONS = 100_000

    fun decrypt(cifradoBase64: String): String {
        val (key, iv) = derivarClaveEIv()
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), IvParameterSpec(iv))
        val textoPlano = cipher.doFinal(Base64.getDecoder().decode(cifradoBase64))
        return String(textoPlano, Charsets.UTF_8)
    }

    /**
     * Rfc2898DeriveBytes.Pbkdf2(passphraseBytes, saltBytes, iterations, SHA256, keyBytes+16)
     * del lado C# — equivalente exacto vía PBKDF2WithHmacSHA256: se derivan
     * (32+16)=48 bytes, los primeros 32 son la clave AES-256, los últimos 16 son el IV.
     * Passphrase/salt son ASCII puro, así que char[] vs UTF-8 bytes no diverge.
     */
    private fun derivarClaveEIv(): Pair<ByteArray, ByteArray> {
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val longitudTotalBits = (KEY_SIZE_BITS + 16 * 8)
        val spec = PBEKeySpec(PASSPHRASE.toCharArray(), SALT.toByteArray(Charsets.UTF_8), ITERATIONS, longitudTotalBits)
        val derivado = factory.generateSecret(spec).encoded
        val key = derivado.copyOfRange(0, KEY_SIZE_BITS / 8)
        val iv = derivado.copyOfRange(KEY_SIZE_BITS / 8, derivado.size)
        return key to iv
    }
}
